src/evaluator.ml

```diff
@@ -1,31 +1,134 @@
 open Ast
+open Common
+let (fresh, reset) =
+   let nxt = ref 0 in
+   let f () = (nxt := !nxt + 1; string_of_int (!nxt)) in
+   let r () = nxt := 0 in
+    (f, r) 
 
+let found_solution = ref false
+                   
+let rec find_vars q  = match q with [] -> []
+                                  | (x::xs) -> (match x with VarExp(v) -> [x] @ (find_vars xs)
+                                                          | ConstExp(c) -> (find_vars xs)
+                                                          | TermExp(s,el) -> (find_vars el) @ (find_vars xs))
+                                            
+(* from https://rosettacode.org/wiki/Remove_duplicate_elements#OCaml *)
```

Comment:
OCaml does support sets: https://caml.inria.fr/pub/docs/manual-ocaml/libref/Set.html

I wonder if it would be simpler to use them.